### PR TITLE
safe GCM nonce construction in TLS 1.2

### LINKED ISSUE #200: Nonces in GCM
Is there a specific problem in TLS?
* Potentially relevant: https://eprint.iacr.org/2015/102.pdf
* Also https://vigilance.fr/vulnerability/TLS-obtaining-data-size-via-HTTPS-Bicycle-18648 ?

### PR DESCRIPTION
**Pull request by thomas-fossati**
Add recommendations for safe GCM nonce construction in TLS 1.2

Fixes #200

### CODE REVIEW

**Comment by yaronf** [comment]
I am good with the text as-is.

The problem with this recommendation is that everything (or almost everything) in RFC7525 can be implemented by an administrator, possibly with the help of a technician armed with Wireshark. This OTOH is an implementation detail, and an admin of a closed source TLS library has no way of knowing if they comply or not. In other words this is just barely in scope.

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 1]
```diff
@@ -160,6 +161,30 @@ informative:
 
   DEP-SSLv3: RFC7568
 
+  Boeck2016:
+    author:
+    - ins: H. Boeck
```
Comment (yaronf) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:166]: Can you try to include the umlaut in the guy's name? See https://github.com/cabo/kramdown-rfc2629/issues/105

Comment (thomas-fossati) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md:166]: sure!

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 2]
```diff
@@ -163,8 +164,8 @@ informative:
 
   Boeck2016:
```
Comment (thomas-fossati) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:165]: I tried to change this occurrence as well but kdrfc didn't like it.

File: BCP195bis/draft-ietf-uta-rfc7525bis.md (hunk 1 above)
Comment (thomas-fossati) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md]: Done in c89df4d

**Comment by yaronf** [comment]
Also, @thomas-fossati please add yourself as an author.

**Comment by thomas-fossati** [comment]
> Also, @thomas-fossati please add yourself as an author.

sure, done in  769d49c

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 3]
```diff
@@ -619,6 +650,27 @@ Host name validation typically applies only to the leaf "end entity" certificate
 {: #sec-aes}
 
 {{rec-cipher}} above recommends the use of the AES-GCM authenticated encryption algorithm. Please refer to Section 11 of {{RFC5246}} for general security considerations when using TLS 1.2, and to Section 6 of {{!RFC5288}} for security considerations that apply specifically to AES-GCM when used with TLS.
+
+### Nonce Reuse in TLS 1.2
+
+The existence of deployed TLS stacks that mistakenly reuse the AES-GCM nonce is
+documented in {{Boeck2016}}, showing there is an actual risk of AES-GCM getting
+implemented in an insecure way, making TLS sessions that use an AES-GCM
```
Comment (thomas-fossati) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:658]:
```suggestion
implemented in an insecure way, and thus making TLS sessions that use an AES-GCM
```

**Comment by stpeter** [comment]
I do see at least one CVE: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2016-0270

**Comment by thomas-fossati** [comment]
> I do see at least one CVE: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2016-0270

thanks!  (AFAIU it's the only one.)

**Comment by stpeter** [comment]
> > I do see at least one CVE: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2016-0270
> 
> thanks! (AFAIU it's the only one.)

Should we mention this CVE in our document?

**Comment by thomas-fossati** [comment]
> > > I do see at least one CVE: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2016-0270
> > 
> > 
> > thanks! (AFAIU it's the only one.)
> 
> Should we mention this CVE in our document?

I think so.

Also,  I need to correct my wrong statement upthread: there are (at least) three other CVEs that apply:
* [CVE-2016-10213](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2016-10213)
* [CVE-2016-10212](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2016-10212)
* [CVE-2017-5933](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2017-5933)

**Review by yaronf — APPROVED** [review]

**Review by stpeter — APPROVED** [review]

**Pull request merged**